src/main/scala/viper/gobra/frontend/Parser.scala

```diff
@@ -442,19 +442,20 @@ object Parser {
       (idnDef <~ "=") ~ typ ^^ { case left ~ right => PTypeAlias(right, left)}
 
     lazy val functionDecl: Parser[PFunctionDecl] =
-      functionSpec ~ ("func" ~> idnDef) ~ signature ~ specOnlyParser(blockWithBodyParameterInfo) ^^ {
-        case spec ~ name ~ sig ~ body =>
-          PFunctionDecl(name, sig._1, sig._2, spec, body)
+      functionSpec() ~ ("func" ~> idnDef) ~ signature ~ specOnlyParser(blockWithBodyParameterInfo) ^^ {
+        case spec ~ name ~ sig ~ body => PFunctionDecl(name, sig._1, sig._2, spec, body)
+      } | functionSpec(withTrusted = true) ~ ("func" ~> idnDef) ~ signature ~ nestedCurlyBracketsConsumer.? ^^ {
+        case spec ~ name ~ sig ~ _ => PFunctionDecl(name, sig._1, sig._2, spec, None)
```

Comment:
I think that there is some repetition here that we can likely avoid. In particular, both branches start by parsing the functionSpec and then obtain the identifier and signature. They only differ in the way that the body is parsed.

Reply:
Something like the following could work:

```scala
    lazy val functionDecl: Parser[PFunctionDecl] =
      for {
        spec <- functionSpec // notice that I removed the param in functionSpec
        name <- ("func" ~> idnDef)
        sig  <- signature
        body <- if (spec.isTrusted) nestedCurlyBracketsConsumer else specOnlyParser(blockWithBodyParameterInfo)
      } yield PFunctionDecl(name, sig._1, sig._2, spec, body)
```

This can likely be improved to look better by using operators from the parsing library instead of the `for` notation

Reply:
This was actually my first attempt of solving this task. I used the `into` operator of the parsing library. The problem with an approach like this is, that if you parse the function spec first, the position of the `PFunctionDecl` node moves down to the `func` key word, implying that the spec doesn't belong to a function declaration. I discussed this with Linard and we came to the conclusion, that this isn't desired behaviour. This is the reason why I used this non-elegant approach

Reply:
`regressions/features/interfaces/proofGenFail1.gobra` would be a test case, where the above described behaviour would cause problems

Reply:
I am not sure if I understand what you mean. Do you mean that the position associated with this AST node is not the intended one? If so, isn't this something that we can overcome by using the `at` method to specify exactly the position of a node? (there are some examples of this method in the Parser class)

Reply:
PS: verifying the file `regressions/features/interfaces/proofGenFail1.gobra` with my proposed code seems to lead to the same exact output as the latest commit of this branch. Is it possible that the erroneous behaviour that you found is caused by some special behaviour of the `into` operator?

Reply:
I didn't know about the `at` operator, this fixes the issue, thanks!

About the problem, the output for the test case is almost the same, the test output should be: 
```
...
[info] 	<proofGenFail1.gobra:24:1> Generated implementation proof (*counter implements interface{ hasNext, next }) failed. Precondition of call to implementation method might not hold. 
[info] Permission to x might not suffice.
[info] 	<proofGenFail1.gobra:30:1> Generated implementation proof (*counter implements interface{ hasNext, next }) failed. Precondition of call to implementation method might not hold. 
[info] Permission to x.f might not suffice.
...
```

without the `at` operator, even with your code, the output looks something like this (there is a difference in the error positions): 
```
...
[info] 	<proofGenFail1.gobra:25:39> Generated implementation proof (*counter implements interface{ hasNext, next }) failed. Precondition of call to implementation method might not hold. 
[info] Permission to x might not suffice.
[info] 	<proofGenFail1.gobra:33:42> Generated implementation proof (*counter implements interface{ hasNext, next }) failed. Precondition of call to implementation method might not hold. 
[info] Permission to x.f might not suffice.
...
```